codec/src/main/java/io/netty/handler/codec/compression/Bzip2BlockCompressor.java

```diff
@@ -197,22 +205,23 @@ boolean write(final int value) {
 
     /**
      * Writes an array to the block.
-     * @param data The array to write
+     * @param buffer The buffer to write
      * @param offset The offset within the input data to write from
      * @param length The number of bytes of input data to write
      * @return The actual number of input bytes written. May be less than the number requested, or
      *         zero if the block is already full
      */
-    int write(final byte[] data, int offset, int length) {
-        int written = 0;
-
-        while (length-- > 0) {
-            if (!write(data[offset++])) {
-                break;
-            }
-            written++;
+    int write(final ByteBuf buffer, int offset, int length) {
+        int index = buffer.forEachByte(offset, length, writeProcessor);
+        if (index == -1) {
+            // Wrote everything
+            return length;
+        }
+        if (offset == index) {
```

Comment:
consider removing this conditional check. arithmetic is more friendly to pipelining than branch statements, and it would be less code. You could even reduce to a tertiary statement:

```java
return index == -1 ? length : index - offset;
```